scripts/pylib/twister/twisterlib/runner.py

```diff
@@ -511,6 +517,47 @@ def process(self, pipeline, done, message, lock, results):
             else:
                 self.cleanup_artifacts()
 
+    def amend_test_instance(self, results):
+        symbol_file = os.path.join(self.build_dir, "zephyr", "zephyr.symbols")
+        if os.path.isfile(symbol_file):
+            logger.debug(f"zephyr.symbols located: {symbol_file}")
+        else:
+            # no zephyr.lst file, cannot do symbol-based amending
+            logger.debug(f"zephyr.symbols NOT located: {symbol_file}")
+            return
+
+        yaml_testsuite_name = self.instance.testsuite.id
+        logger.debug(f"amending yaml test suite: {yaml_testsuite_name}")
+
+        with open(symbol_file, 'r') as fp:
+            symbols = fp.read()
+            logger.debug(f"test instance {self.instance.name} has {len(self.instance.testcases)} cases.")
+            for testcase in self.instance.testcases:
```

Comment:
That is where I am getting confused, you are checking against what was parsed earlier using regex, but the intent is to replace that and just use the input from the symbols file

Reply:
I was planning to solely rely on the zephyr.symbols file. 

But it cannot handle below scenario:

https://github.com/zephyrproject-rtos/zephyr/blob/b2f80b0e16363021bc1fd65884b11830fa869870/tests/lib/cmsis_dsp/matrix/src/binary_q7.c#L119-L126

The `test_matrix_binary_q7()` and `test_op2_arm_mat_mult_q7()` both have `test_` prefix but only the latter is a valid unit test. I cannot tell them apart just with the `zephyr.symbols` file.  Even `test_main()` has a `test_` prefix.

Though the symbol file does show some subtle difference between them as shown below. The `T` means global text section and `t` means local text section.
```
00000ca8 T test_matrix_binary_q7
00000a0c t test_op2_arm_mat_mult_q7
```
But that's not always true because in our code, a unit test function can be declared with or without `static` keyword.
For example below one is a unit test without `static`:

https://github.com/zephyrproject-rtos/zephyr/blob/b2f80b0e16363021bc1fd65884b11830fa869870/tests/crypto/tinycrypt/src/aes.c#L71

So, we have to keep the old regex-based method for the sake of old-style tests.

If we add another test collection method for the new ztest fx **right now**, that means we may need to tell which method to use when scanning for tests.  We may make a choice by checking for `CONFIG_NEW_ZTEST` in the final `.config`.

But the regex-based collection happens **before** build, while symbol-based collection happens **after** build. Currently the regex-based collection **can** collect for both styles. While the symbol-based collection can only cover the new style. When the time comes to carry out symbol-based collection, all tests have **already** been collected, though with some falseness. So I choose to amend rather than re-collect the new-style tests.

Actually I consider this PR as just **phase 1**, which aims to remove false test suite/test function combinations such as in cmsis tests.

In future, when we fully migrate to the new ztest fx. A **phase 2** PR can be done to solely rely on the `zephyr.symbols`. By then, we will be able to readily use the `z_ztest_unit_test` prefix to find unit tests and `z_ztest_test_node` to find test suite.

Reply:
> So, we have to keep the old regex-based method for the sake of old-style tests.

I am not suggesting it should be removed now.

The tests you mentioned above were not converted to new ztest API yet and those still need to be parsed using regex, you should not try and get the tests from binaries of tests that were not converted.

> If we add another test collection method for the new ztest fx right now, that means we may need to tell which method to use when scanning for tests.

In that case you will be using the symbols file itself, if you are using some regex to collect the data (using new ztest api) then you will get no matches with old tests and can skip in that case.

> So I choose to amend rather than re-collect the new-style tests.

I would just override 

> Actually I consider this PR as just phase 1, which aims to remove false test suite/test function combinations such as in cmsis tests.

Yes, this is phase 1, I agree.

Reply:
> In that case you will be using the symbols file itself, if you are using some regex to collect the data (using new ztest api) then you will get no matches with old tests and can skip in that case.

I think my current way to utilize the `zephyr.symbols` for amending can benefit both old and new styles tests.
Maybe this is an optimum result for phase 1?

> I would just override

I feel I am kind of overriding the regex-parsing results now. Though I call it amend. 
Do you mean besides amending the invalid ones from regex parsing, I should also **re-add** the valid ones found with symbols, in spite of that they have already been found through regex parsing?

Is that really necessary since I saw the regex-parsing logic which caters for the new ztest:

https://github.com/zephyrproject-rtos/zephyr/blob/b2f80b0e16363021bc1fd65884b11830fa869870/scripts/pylib/twister/twisterlib/testsuite.py#L76-L78

Reply:
@nashif  Just added logic to override the old regex-based test case collection result. 
And it should also work when no regex-based test collection exists. Will test more on that.

Reply:
@nashif  Updated the commit message to reflect what's going on.